src/main/java/com/mraof/minestuck/item/StructureScannerItem.java

```diff
@@ -2,84 +2,180 @@
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.tags.TagKey;
 import net.minecraft.world.InteractionHand;
 import net.minecraft.world.InteractionResultHolder;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.levelgen.structure.Structure;
 
+import javax.annotation.Nullable;
 import java.util.function.Supplier;
 
+/**
+ * Generic class for scanners which locates the nearest structure of a certain type in the player's current dimension.
+ * Takes an optional fuel item, which will be consumed every tick until the device deactivates.
+ * Toggles on or off with right-click, or when fuel runs out.
+ */
+
 public class StructureScannerItem extends Item
 {
 	private final TagKey<Structure> structure;
+	@Nullable
 	private final Supplier<Item> fuelItem;
 	
-	public StructureScannerItem(Properties properties, TagKey<Structure> structure, Supplier<Item> fuelItem)
+	public StructureScannerItem(Properties properties, TagKey<Structure> structure, @Nullable Supplier<Item> fuelItem)
 	{
 		super(properties);
 		this.structure = structure;
 		this.fuelItem = fuelItem;
 	}
 	
 	@Override
-	public InteractionResultHolder<ItemStack> use(Level levelIn, Player playerIn, InteractionHand handIn)
+	public InteractionResultHolder<ItemStack> use(Level pLevel, Player pPlayer, InteractionHand pUsedHand)
 	{
-		levelIn.playSound(playerIn, playerIn.getX(), playerIn.getY(), playerIn.getZ(), SoundEvents.UI_BUTTON_CLICK.value(), SoundSource.AMBIENT, 0.8F, 1.3F);
-		
-		ItemStack item = playerIn.getItemInHand(handIn);
-		boolean foundItem = false;
+		ItemStack stack = pPlayer.getItemInHand(pUsedHand);
 		
-		if(fuelItem != null)
+		if(checkFuelNeeded(pPlayer, pLevel) || pPlayer.isCreative())
 		{
-			ItemStack fuelStack = new ItemStack(fuelItem.get());
+			if (stack.getDamageValue() > 0){
+				resetCharge(stack);
+			}
+			stack.getOrCreateTag().putBoolean("Powered", true);
+			pLevel.playSound(pPlayer, pPlayer.getX(), pPlayer.getY(), pPlayer.getZ(), SoundEvents.UI_BUTTON_CLICK.get(), SoundSource.AMBIENT, 0.8F, 1.3F);
 			
-			for(ItemStack invItem : playerIn.getInventory().items)
+			if (!pLevel.isClientSide)
 			{
-				if(ItemStack.isSameItem(invItem, fuelStack))
-				{
-					foundItem = true;
-					if(!levelIn.isClientSide && levelIn.getRandom().nextFloat() >= 0.95F)
-					{
-						invItem.shrink(1);
-						levelIn.playSound(playerIn, playerIn.getX(), playerIn.getY(), playerIn.getZ(), SoundEvents.BLAZE_SHOOT, SoundSource.AMBIENT, 0.4F, 2F);
-					}
-					
-					break;
-				}
+				MutableComponent message = Component.translatable("message.temple_scanner.on");
+				pPlayer.sendSystemMessage(message.withStyle(ChatFormatting.DARK_GREEN));
+			}
+		}
+		return InteractionResultHolder.success(stack);
```

Comment:
Would it be reasonable to return a success result when it was successfully turned on, and failure if fuel was missing?